### PR TITLE
#42 Autoslaughter

### LINKED ISSUE #42: Auto slaughter over X
For breeding farms auto-slaughter, oldest first when count over X

Needs:
* a slider for limit to slaughter beyond
* a tick box for slaughter on
* never slaughter the last of any gender
* a jobgiver / workgiver to auto-slaughter if should be slaughtered OR perhaps the building auto-slaughters and the next produce tick just kills the pawn and dumps the meat.

### PR DESCRIPTION
**Pull request by keyz182**
Closes #42

### CODE REVIEW

**Comment by keyz182** [comment]
Recreated a cut down version of the normal autoslaughter manager, inside a tab. Omits the pregnant and bonded columns.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmNutrition.cs [hunk 1]
```diff
@@ -55,6 +55,7 @@ public override void PostExposeData()
         {
             base.PostExposeData();
             Scribe_Values.Look(ref _storedNutrition, "storedNutrition");
+            Scribe_References.Look(ref AlternativeStore, "AlternativeStore");
```
Comment (keyz182) [inline 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmNutrition.cs:58]: This also fixes a bug I found where nutrition in pipe based systems wasn't being persisted.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmBreeder.cs [hunk 2]
```diff
@@ -38,13 +90,62 @@ public override void CompTick()
...
-                        into def
-                        select def;
 
+                    foreach (var type in types)
+                    {
+                        AutoSlaughterConfig config = AutoSlaughterSettings.FirstOrDefault(s => s.animal == type.Key.race);
+
+                        if (config != null)
+                        {
+                            // Only slaughter one per cycle
```
Comment (feldoh) [inline 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmBreeder.cs:107]: might be overkill but this is so duplicatey I'd consider extracting it into a CullOldest function then you only need to pass in the lists or even just the whole types thing. As you're only culling one you don't need to pre-make all of them either just one at a time; maybe have a culledThisCycle bool then it's just something like

```csharp
bool culledThisCycle = false;
if (!culledThisCycle) culledThisCycle = cullOldestOverLimit(config.maxMales, type.Where(p => p.gender == Gender.Male && p.ageTracker.Adult).ToList());
if (!culledThisCycle) culledThisCycle = cullOldestOverLimit(config.maxFemales, type.Where(p => p.gender == Gender.Female && p.ageTracker.Adult).ToList());
if (!culledThisCycle) culledThisCycle = cullOldestOverLimit(config.maxMales, type.Where(p => p.gender == Gender.Male && !p.ageTracker.Adult).ToList());
if (!culledThisCycle) culledThisCycle = cullOldestOverLimit(config.maxFemales, type.Where(p => p.gender == Gender.Female && !p.ageTracker.Adult).ToList());
```

your CountForType util even already basically does this right

If you really wanted to go ham you could even roll it all into one iteration, just groupBy ageTracker and gender. Then a Find using the cull function which returns true if culled so it'll just do one iteration then cull one from one group and stop. Might be a bit harder to read though so your call.

In general there's way too much going on in this function now too, we're better off extracting the steps to like TryCull, TryBreed etc

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmNutrition.cs (hunk 1 above)
Comment (feldoh) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmNutrition.cs:58]: awesome spot.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs [hunk 3]
```diff
@@ -24,9 +24,35 @@ public class CompFarmStorage : CompPawnStorage
...
+        public IEnumerable<PawnKindDef> HeldPawnTypes => this.innerContainer.innerList.Select(p => p.kindDef).Distinct();
+
+        public Dialog_AutoSlaughter.AnimalCountRecord CountForType(ThingDef def)
+        {
+            List<Pawn> pawns = innerContainer.innerList.Where(p => p.kindDef.race == def).ToList();
+            int total = pawns.Count;
+            int male = pawns.Count(p => p.gender == Gender.Male && p.ageTracker.Adult);
+            int maleYoung = pawns.Count(p => p.gender == Gender.Male && !p.ageTracker.Adult);
+            int female = pawns.Count(p => p.gender == Gender.Female && p.ageTracker.Adult);
+            int femaleYoung = pawns.Count(p => p.gender == Gender.Female && !p.ageTracker.Adult);
```
Comment (feldoh) [inline 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs:36]: can this not be replaced by your CountForType util

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs [hunk 4]
```diff
@@ -24,9 +24,35 @@ public class CompFarmStorage : CompPawnStorage
...
+            if(despawnFirst)
+                pawn.DeSpawn();
+
+            if (innerContainer.Count >= MaxStoredPawns())
+            {
+                Messages.Message("PS_StorageFull".Translate(parent.LabelCap, pawn.LabelCap), (Thing) pawn, MessageTypeDefOf.NeutralEvent);
+
+                PawnComponentsUtility.AddComponentsForSpawn(pawn);
+                compAssignable?.TryUnassignPawn(pawn);
+                GenDrop.TryDropSpawn(pawn, parent.Position, parent.Map, ThingPlaceMode.Near, out Thing _);
```
Comment (feldoh) [inline 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs:52]: I think this has a potential bug; if despawnFirst is false and the storage is full it'll dupe the pawn.
Why would you ever not despawn first?

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Interfaces/IBreederParent.cs [hunk 5]
```diff
@@ -6,8 +6,13 @@ namespace PawnStorages.Farm.Interfaces
     public interface IBreederParent : IActive
     {
         public List<Pawn> BreedablePawns { get; }
+        public List<Pawn> AllPawns { get; }
         public int TickInterval { get; }
 
         public void Notify_PawnBorn(Pawn newPawn);
+
+        public void Release(Pawn pawn);
```
Comment (feldoh) [inline 1.5/Source/PawnStorages/PawnStorages/Farm/Interfaces/IBreederParent.cs:14]: if you call this ReleasePawn you avoid the earlier duplication

**Review by feldoh — CHANGES_REQUESTED** [review]

File: 1.5/Source/PawnStorages/PawnStorages/Farm/AutoSlaughterGizmo.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    public const float CostPreviewFadeOut = 0.6f;
+    private static readonly Color PainBoostColor = new Color(0.2f, 0.65f, 0.35f);
+    private static readonly Texture2D EntropyBarTex = SolidColorMaterials.NewSolidColorTexture(new Color(0.46f, 0.34f, 0.35f));
+    private static readonly Texture2D EntropyBarTexAdd = SolidColorMaterials.NewSolidColorTexture(new Color(0.78f, 0.72f, 0.66f));
+    private static readonly Texture2D OverLimitBarTex = SolidColorMaterials.NewSolidColorTexture(new Color(0.75f, 0.2f, 0.15f));
+    private static readonly Texture2D PsyfocusBarTex = SolidColorMaterials.NewSolidColorTexture(new Color(0.34f, 0.42f, 0.43f));
+    private static readonly Texture2D PsyfocusBarTexReduce = SolidColorMaterials.NewSolidColorTexture(new Color(0.65f, 0.83f, 0.83f));
+    private static readonly Texture2D PsyfocusBarHighlightTex = SolidColorMaterials.NewSolidColorTexture(new Color(0.43f, 0.54f, 0.55f));
+    private static readonly Texture2D EmptyBarTex = SolidColorMaterials.NewSolidColorTexture(new Color(0.03f, 0.035f, 0.05f));
+    private static readonly Texture2D PsyfocusTargetTex = SolidColorMaterials.NewSolidColorTexture(new Color(0.74f, 0.97f, 0.8f));
```
[the remark this answers is not in the document]
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/AutoSlaughterGizmo.cs]: Oops! Deleted.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/AutoSlaughterGizmo.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+      if (Mathf.CeilToInt(this.lastTargetValue) != this.TargetValue)
+      {
+          breeder.AutoSlaughterTarget = TargetValue;
+      }
+
+      // Rect tooltipRect = insetRect;
+      // tooltipRect.y += 38f;
+      // tooltipRect.width = 175f;
+      // tooltipRect.height = 38f;
+      // TooltipHandler.TipRegion(tooltipRect, (Func<string>) (() => this.tracker.PsyfocusTipString(this.selectedPsyfocusTarget)), Gen.HashCombineInt(this.tracker.GetHashCode(), 133873));
```
[the remark this answers is not in the document]
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/AutoSlaughterGizmo.cs]: Gone

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Building_PSFarm.cs [hunk 8]
```diff
@@ -132,6 +137,15 @@ public void ReleasePawn(Pawn pawn)
 
         public void Notify_NutritionNotEmpty() => NutritionAvailable = true;
 
+        public List<Pawn> AllPawns
```
[the remark this answers is not in the document]
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Building_PSFarm.cs]: Agreed

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Building_PSFarm.cs [hunk 9]
```diff
@@ -124,6 +124,11 @@ public void ReleasePawn(Pawn pawn)
             pawnStorage.ReleaseSingle(Map, pawn, true, true);
         }
 
+        public void Release(Pawn pawn)
+        {
+            ReleasePawn(pawn);
+        }
+
```
[the remark this answers is not in the document]
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Building_PSFarm.cs]: Yup, added a new method to the interface without thinking, and then just created this. Tweaked the interface and removed this.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmBreeder.cs [hunk 10]
```diff
@@ -1,16 +1,38 @@
...
     public class CompFarmBreeder : ThingComp
     {
+        public int AutoSlaughterTarget = 0;
         public IBreederParent Parent => parent as IBreederParent;
 
+        public List<AutoSlaughterConfig> AutoSlaughterSettings = new List<AutoSlaughterConfig>();
+        private void TryPopulateMissingAnimals()
+        {
+            HashSet<ThingDef> hashSet = new HashSet<ThingDef>();
+            hashSet.AddRange(AutoSlaughterSettings.Select(c => c.animal));
```
[the remark this answers is not in the document]
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmBreeder.cs]: I did do a dict originally. Switched to this when trying to follow the auto slaughter manager stuff. Changed back to a dict now though.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmBreeder.cs (hunk 2 above)
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmBreeder.cs]: Completely refactored based roughly on what you suggested. Likely needs a good looking over.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs (hunk 3 above)
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs:36]: This _is_ the `CountForType` util :)

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs (hunk 4 above)
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Comps/CompFarmStorage.cs:52]: You wouldn't despawn first in the case where we've generated a new pawn from breeding, but not spawned it yet. 
In this case, when a pawn is "born" it's generated, but not spawned, then passed to the `Notify_PawnBorn` method.
This passes it to `StorePawn` which either stores it, or spawns it. That's the only codepath where `despawnFirst` will be false.

File: 1.5/Source/PawnStorages/PawnStorages/Farm/Interfaces/IBreederParent.cs (hunk 5 above)
Comment (keyz182) [reply 1.5/Source/PawnStorages/PawnStorages/Farm/Interfaces/IBreederParent.cs]: Yup, done.

**Comment by feldoh** [comment]
I think the more reliable option is to just check the pawn is spawned
before de spawning. That accounts for never spawned pawns as well as any
other cases where the pawn isn’t spawned for whatever reason

**Comment by feldoh** [comment]
Whoops, that’ll teach me for copy pasting a comment to all the similar code
blocks XD

**Pull request merged**